protonmail-desktop: init at 1.0.1

rsniezek opened the pull request:
Add a package for recently shipped protonmail desktop client in beta version.

The package definition HEAVILY based on brave browser and needs further cleanup as most of the deps are most likely unneeded. But the app builds, starts and even reads/sends mails.

Testing and review is very welcomed.

## Description of changes
closes https://github.com/NixOS/nixpkgs/issues/296107

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

rsniezek commented:
I think I removed most stuff that were not really nescessary.

dotlambda commented on pkgs/by-name/pr/protonmail-desktop/package.nix:
```diff
@@ -0,0 +1,95 @@
+  binpath = makeBinPath deps;
+in
+stdenv.mkDerivation rec {
+  pname = "ProtonMail-desktop";
```
```suggestion
  pname = "protonmail-desktop";
```

rsniezek replied:
changed to proton-mail so it's usable in the URL and becaus that's the main program delivered. No conflicts with other packages. Maintainer change extracted to separate commit.

dotlambda commented on pkgs/by-name/pr/protonmail-desktop/package.nix:
```diff
@@ -0,0 +1,95 @@
+
+  src = fetchurl {
+    url = "${baseUrl}/${pname}-${version}.deb";
+    sha256 = "sha256-3FRm02Ew9GuGMXFAkwSL+FbeLyaI1SmpDU3A5eJh6TU=";
```
```suggestion
    hash = "sha256-3FRm02Ew9GuGMXFAkwSL+FbeLyaI1SmpDU3A5eJh6TU=";
```

rsniezek replied:
done

dotlambda commented on pkgs/by-name/pr/protonmail-desktop/package.nix:
```diff
@@ -0,0 +1,95 @@
+stdenv.mkDerivation rec {
+  pname = "ProtonMail-desktop";
+  baseUrl = "https://proton.me/download/mail/linux/";
+  version = "beta";
```
This sounds like they would just upload a new deb without updating the version, which would invalidate the hash.

fin444 replied:
Can also be gotten from https://github.com/ProtonMail/inbox-desktop/releases, SHA256 matches exactly with `proton-mail_1.0.0_amd64.deb`

rsniezek replied:
Changed the url to github so it's versioned and guarded from this kind of modification by the provider.

dotlambda commented on pkgs/by-name/pr/protonmail-desktop/package.nix:
```diff
@@ -0,0 +1,95 @@
+  meta = with lib; {
+    description = "Desktop application for Proton Mail";
+    homepage = "https://www.proton.me/mail";
+    license = licenses.free;
```
Which one?

fin444 replied:
GPLv3 maybe? https://github.com/ProtonMail/inbox-desktop
But that repo doesn't actually have the source code

SebTM replied:
Hey, from what I know the license is MIT :)

dotlambda replied:
Can you send a link confirming that?

urandom2 replied:
this one?: https://github.com/ProtonMail/inbox-desktop/blob/main/LICENSE

dotlambda replied:
No. Something can't be licensed under GPL without the code being public.

urandom2 replied:
iiuc, the code does not need to be public, you just need the ability to request it

rsniezek replied:
I've reached out to proton team with question about license and source code availability.

SebTM replied:
Hey, to maybe clear this up a bit: I'm part/employee of Proton (not that particular team) and already reached out to them to clarify - what I can already say for sure it's planned as with all our clients to open-source the source-code :v:

rsniezek replied:
changed to gpl3Plus as this seems to be most accurate. @SebTM in case you get any conflicting info from the team, feel free to update.

dotlambda reviewed (commented):
`meta.mainProgram` and `meta.sourceProvenance` are missing

Please fix the commit message and move the change to the maintainer list to another commit.

rsniezek commented:
Thanks for the inputs, will send update either later today or Monday

SebTM commented:
Hey, I did the same but wasn't finished to publish it yet because of testing:
```
{ stdenv
, lib
, fetchurl
, dpkg
, electron
, makeWrapper
}: let
  pname = "protonmail-desktop";
  binName = "proton-mail";

in stdenv.mkDerivation {
  inherit pname;
  version = "1.0.0-beta";

  src = fetchurl {
    url = "https://proton.me/download/mail/linux/ProtonMail-desktop-beta.deb";
    hash = "sha256-3FRm02Ew9GuGMXFAkwSL+FbeLyaI1SmpDU3A5eJh6TU=";
  };

  dontConfigure = true;
  dontBuild = true;

  nativeBuildInputs = [ makeWrapper dpkg ];

  unpackPhase = "dpkg-deb --fsys-tarfile $src | tar -x --no-same-permissions --no-same-owner";

  installPhase = ''
    runHook preInstall

    mkdir -p $out/bin
    cp -r usr/share/ $out/
    cp -r usr/lib/proton-mail/resources/app.asar $out/share/

    makeWrapper ${electron}/bin/electron $out/bin/${binName} \
      --add-flags $out/share/app.asar \
      --add-flags "\''${NIXOS_OZONE_WL:+\''${WAYLAND_DISPLAY:+--ozone-platform-hint=auto --enable-features=WaylandWindowDecorations}}" \
      --set-default ELECTRON_FORCE_IS_PACKAGED 1 \
      --set-default ELECTRON_IS_DEV 0 \
      --inherit-argv0

    runHook postInstall
  '';

  meta = with lib; {
    description = "Proton Mail Desktop-App for Linux";
    homepage = "https://github.com/ProtonMail/desktop";
    license = licenses.gpl3Plus;
    platforms = platforms.linux;
    mainProgram = binName;
    maintainers = with maintainers; [ sebtm ];
  };
}
```
This is what I ended up with, not problems so far but only used a few hours.

SebTM commented on pkgs/by-name/pr/protonmail-desktop/package.nix:
```diff
@@ -0,0 +1,95 @@
+  '';
+
+  preFixup = ''
+    # Add command line args to wrapGApp.
```
Do we need gappsWrapper here makeWrapper worked fine for me?

rsniezek replied:
Seems not. I changed the package to be a mix of mine and yours and it is now much simpler and still working.
I tested on plasma6/wayland with both xwayland and native via ozone and it seems to be working fine.

SebTM commented on pkgs/by-name/pr/protonmail-desktop/package.nix:
```diff
@@ -0,0 +1,95 @@
+    gappsWrapperArgs+=(
+      --prefix LD_LIBRARY_PATH : ${rpath}
+      --prefix PATH : ${binpath}
+      --suffix PATH : ${lib.makeBinPath [ xdg-utils coreutils ]}
```
Why did you need them? I didn't see any crashes regarding missing dependencies

rsniezek replied:
Seems those were not needed after all. makeWrapper is enough.

SebTM commented:
```
      --add-flags "\''${NIXOS_OZONE_WL:+\''${WAYLAND_DISPLAY:+--ozone-platform-hint=auto --enable-features=WaylandWindowDecorations}}" \
```
Would be important to run it native on wayland if the env-variable (from the wiki) is set

---

```
      --set-default ELECTRON_FORCE_IS_PACKAGED 1 \
      --set-default ELECTRON_IS_DEV 0 \
```
I had to set (one or both) to use the right servers (prod) for the app

SebTM commented:
P.S. Feel free to add me as maintainer as well in the updated-version :v:

rsniezek commented:
> ```
> 
> ```
> 
> 
>     
> 
>   
> 
> Would be important to run it native on wayland if the env-variable (from the wiki) is set

With the variable set I get wayland decorations on the window, without the variable i get XWayland decorations (plasma6/wayland). So the flags need to be added.

rsniezek commented:
> I had to set (one or both) to use the right servers (prod) for the app

How do you determine that?

rsniezek committed:
protonmail-desktop: init at 1.0.1

Add package for proton mail desktop app in the beta version.

Apply suggestions from code review

Apply batch of suggestions from CR by @dotlambda.

Co-authored-by: Robert Schütz <github@dotlambda.de>

rsniezek committed:
maintainers: add rsniezek

Adding myself as a maintainer so that I can add myself in packages I work on.

dotlambda commented on pkgs/by-name/pr/proton-mail/package.nix:
```diff
@@ -0,0 +1,54 @@
+stdenv.mkDerivation rec {
+  pname = "proton-mail";
+  versionBase = "1.0.0";
+  versionSuffix = "beta";
```
Why do we need a suffix?

rsniezek replied:
we don't need it. the verison of the .deb is 1.0.0 but the "beta" is implied in the release notes and in the description. I thought it would be fair to the users to know that this is in fact still in beta. But if you don't think it's needed, I can make this just version = "1.0.0". No strong feelings here

dotlambda replied:
We should use the version used by upstream, not come up with our own versioning scheme.

dotlambda commented on pkgs/by-name/pr/proton-mail/package.nix:
```diff
@@ -0,0 +1,54 @@
+  version = "${versionBase}-${versionSuffix}";
+
+  src = fetchurl {
+    url = "https://github.com/ProtonMail/inbox-desktop/releases/download/v${versionBase}/${pname}_${versionBase}_amd64.deb";
```
```suggestion
    url = "https://github.com/ProtonMail/inbox-desktop/releases/download/v${versionBase}/protpn-mail_${versionBase}_amd64.deb";
```
See #277994. Also applies to other occurrences.

rsniezek replied:
used helper (non pname) variable isntead to minimize the risk of this divering in different lines

dotlambda commented on pkgs/by-name/pr/proton-mail/package.nix:
```diff
@@ -0,0 +1,54 @@
+
+  nativeBuildInputs = [ dpkg makeWrapper ];
+
+  unpackPhase = "dpkg-deb --fsys-tarfile $src | tar -x --no-same-permissions --no-same-owner";
```
```suggestion
  unpackPhase = ''
    dpkg-deb --fsys-tarfile $src | tar -x --no-same-permissions --no-same-owner
  '';
```

rsniezek replied:
ack

dotlambda commented on pkgs/by-name/pr/proton-mail/package.nix:
```diff
@@ -0,0 +1,54 @@
+    mkdir -p $out
+    cp -r usr/share/ $out/
+    cp -r usr/lib/proton-mail/resources/app.asar $out/share/
+    rm -rf usr
```
```suggestion
```

rsniezek replied:
ack

dotlambda commented on pkgs/by-name/pr/proton-mail/package.nix:
```diff
@@ -0,0 +1,54 @@
+  '';
+
+  preFixup = ''
+    makeWrapper ${electron}/bin/electron $out/bin/${pname} \
```
```suggestion
    makeWrapper ${lib.getExe electron} $out/bin/proton-mail \
```

rsniezek replied:
ack

dotlambda commented on pkgs/by-name/pr/proton-mail/package.nix:
```diff
@@ -0,0 +1,54 @@
+  meta = with lib; {
+    description = "Desktop application for Mail and Calendar, made with Electron";
+    homepage = "https://github.com/ProtonMail/inbox-desktop";
+    license = licenses.gpl3Plus;
```
Is it currently possible to obtain the source code under this license?

rsniezek replied:
The source is not hosted anywhere but GPL3 doesn't require that. The distributor of the SW must make it available to the clients, not the general public. There are open issues on Proton's GH with the question about the source code but no replies yet. Previous versions of the desktop app had sources released AFAIK. The licence is taken from the proton's github.

dotlambda replied:
I know. Those issues show that it's not *actually* possible to obtain the source code. Thus I don't believe this license declaration.

rsniezek replied:
I agree, I'm also skeptical. But it's not my place to declare other licence than the upstream. So unless proton team gives me a response with different license, this has to stay for now.

dotlambda replied:
> So unless proton team gives me a response with different license, this has to stay for now.

Why? We can mark it unfree until it's possible to obtain the source.

dotlambda commented on pkgs/by-name/pr/proton-mail/package.nix:
```diff
@@ -0,0 +1,54 @@
+}
+:
+stdenv.mkDerivation rec {
+  pname = "proton-mail";
```
We should consider calling it the same as other distros: https://repology.org/project/protonmail-desktop/versions

rsniezek replied:
ack

dotlambda reviewed (commented):
First commit should have prefix `maintainers:`.

dotlambda commented on pkgs/by-name/pr/protonmail-desktop/package.nix:
```diff
@@ -0,0 +1,54 @@
+, makeWrapper
+, dpkg
+, electron
+}
+:
+stdenv.mkDerivation rec {
```
```suggestion
}:

let
  mainProgram = "proton-mail";
in stdenv.mkDerivation rec {
```

dotlambda commented on pkgs/by-name/pr/protonmail-desktop/package.nix:
```diff
@@ -0,0 +1,54 @@
+stdenv.mkDerivation rec {
+  pname = "protonmail-desktop";
+  version = "1.0.0";
+  binName = "proton-mail";
```
```suggestion
```

dotlambda commented on pkgs/by-name/pr/protonmail-desktop/package.nix:
```diff
@@ -0,0 +1,54 @@
+  '';
+
+  preFixup = ''
+    makeWrapper ${lib.getExe electron} $out/bin/${binName} \
```
```suggestion
    makeWrapper ${lib.getExe electron} $out/bin/${mainProgram} \
```

SebTM commented:
> > I had to set (one or both) to use the right servers (prod) for the app
> 
> How do you determine that?

Without both I get a white window and 

```
[2024-03-18 00:34:50.430] [error] getSessionID TypeError: Invalid URL
    at new URL (node:internal/url:775:36)
    at t.getSessionID (/nix/store/sk9qmbzdw5y9mbi9rpf2rijz3arj475n-protonmail-desktop-1.0.0-beta/share/app.asar/.webpack/main/index.js:8:233403)
    at WebContents.<anonymous> (/nix/store/sk9qmbzdw5y9mbi9rpf2rijz3arj475n-protonmail-desktop-1.0.0-beta/share/app.asar/.webpack/main/index.js:8:215442)
    at WebContents.emit (node:events:514:28)
```

in the log, with only "ELECTRON_FORCE_IS_PACKAGED" with value "1" it seems to still work (quick test, login, sent one test-mail) works.

One issue I found in the test is that the window is somehow cut at the bottom which e.g. does not let you reach the send button easily. Will confirm if that's a gnome/packaging/general issue:

![Screenshot from 2024-03-19 19-47-55](https://github.com/NixOS/nixpkgs/assets/17243347/82ab7e29-622c-4e9e-b015-e3fdc245e45f)

SebTM reviewed (approved):
Good first start, using it like this since a bit 👍🏻

dotlambda commented:
Please squash.

rsniezek commented:
So what's next? I don't have access to merge this. Anything else needs to be done?

MikaelFangel commented:
@rsniezek - Usually it will be open for a little while to allow others to review the PR and then it will be merged by one with the proper rights.

fin444 commented:
[v1.0.1 has been released](https://github.com/ProtonMail/inbox-desktop/releases/tag/v1.0.1)

rsniezek commented:
> [v1.0.1 has been released](https://github.com/ProtonMail/inbox-desktop/releases/tag/v1.0.1)

no Linux changes in the release notes but updated regardless

fin444 commented:
> no Linux changes in the release notes but updated regardless

Yeah but it has an "updates available" popup on launch

fin444 reviewed (approved):
LGTM

dotlambda commented:
Why am I the author of the commit?

Also, is [desktop-release-1.0.1.zip](https://github.com/ProtonMail/inbox-desktop/releases/download/v1.0.1/desktop-release-1.0.1.zip) the source code? Can someone try to build this from source?

SebTM reviewed (approved):
Works for me without changes - regarding source-build I already gave it a try and will summarize it but would prefer to merge this first and then iterate over the source-build.

Update see [discourse](https://discourse.nixos.org/t/build-support-electron-forge-based-applications-e-g-protonmail-desktop-source-build/41974/1) or [here](https://github.com/NixOS/nixpkgs/pull/297822) for my approach ✌🏻

rsniezek commented:
> Why am I the author of the commit?

I applied your suggestions and GH added the co-authored line automatically. Should I remove it?

foolnotion commented:
Works fine for me too, looking forward to see this merged.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/build-support-electron-forge-based-applications-e-g-protonmail-desktop-source-build/41974/1

wegank commented on pkgs/by-name/pr/protonmail-desktop/package.nix:
```diff
@@ -0,0 +1,55 @@
+    dpkg-deb --fsys-tarfile $src | tar -x --no-same-permissions --no-same-owner
+  '';
+
+  installPhase = ''
+    mkdir -p $out
```
```suggestion
  installPhase = ''
    runHook preInstall
    mkdir -p $out
```

wegank commented on pkgs/by-name/pr/protonmail-desktop/package.nix:
```diff
@@ -0,0 +1,55 @@
+
+  nativeBuildInputs = [ dpkg makeWrapper ];
+
+  unpackPhase = ''
+    dpkg-deb --fsys-tarfile $src | tar -x --no-same-permissions --no-same-owner
+  '';
```
Also `runHook {pre,post}Unpack` here (or use `unpackCmd` instead?)

rsniezek replied:
removed completely, the dpkg's built-in setup-hook works for me.

SebTM reviewed (approved):
Still works well for me 👍🏻 

Result of `nixpkgs-review pr 296127` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>protonmail-desktop</li>
  </ul>
</details>
